### PR TITLE
Refactor to pep 0585

### PR DESCRIPTION
**Pull request by henrikstranneheim**
## Description

Refactor to pep 0585

### Changed

- Refactor to pep 0585
- Removed unused Session imports

### How to prepare for test
- [ ] ssh to hasta (depending on type of change)
- [ ] activate stage: `us`
- [ ] request trailblazer-stage on hasta: `paxa`
- [ ] install on stage:
    ```Shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_trailblazer -t trailblazer -b refactor-to-pep-0585-a
    ```
- [ ] ssh to clinical-db (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-trailblazer-ui-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [ ] login to ...
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: tests/apps/slurm/test_api.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-from typing import Dict, List, Optional
+from typing import Optional
```
Comment (diitaz93) [inline tests/apps/slurm/test_api.py:1]: I think we can include in this PR changing the occurrences of `Optional[type]` for `type | None`

Comment (diitaz93) [reply tests/apps/slurm/test_api.py:1]: there is one occurrence in this file in line 137

File: tests/cli/test_cli_core.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import subprocess
 from datetime import datetime
-from typing import Dict, Optional
+from typing import Optional
```
Comment (diitaz93) [inline tests/cli/test_cli_core.py:3]: Also in this file some type hints that could be updated to python 3.11 stardards

File: tests/store/crud/test_read.py [hunk 3]
```diff
@@ -157,7 +157,7 @@ def test_get_analyses_for_case(analysis_store: MockStore, case_id: str):
     # GIVEN a store with an analysis
 
     # WHEN accessing it by case id
-    analyses: Optional[List[Analysis]] = analysis_store.get_analyses_for_case(case_id=case_id)
+    analyses: Optional[list[Analysis]] = analysis_store.get_analyses_for_case(case_id=case_id)
```
Comment (diitaz93) [inline tests/store/crud/test_read.py:160]:
```suggestion
    analyses: list[Analysis] | None = analysis_store.get_analyses_for_case(case_id=case_id)
```

File: tests/store/crud/test_read.py [hunk 4]
```diff
@@ -168,7 +168,7 @@ def test_get_analyses_with_statuses(analysis_store: MockStore):
     # GIVEN a store with an analysis
 
     # WHEN accessing an analysis by statuses
-    analyses: Optional[List[Analysis]] = analysis_store.get_analyses_with_statuses(
+    analyses: Optional[list[Analysis]] = analysis_store.get_analyses_with_statuses(
```
Comment (diitaz93) [inline tests/store/crud/test_read.py:171]:
```suggestion
    analyses: list[Analysis] | None = analysis_store.get_analyses_with_statuses(
```

File: tests/store/crud/test_update.py [hunk 5]
```diff
@@ -180,15 +180,15 @@ def test_update_analysis_jobs_from_slurm_jobs(analysis_store: MockStore, squeue_
 def test_update_case_analyses_as_deleted(analysis_store: MockStore, ongoing_analysis_case_id: str):
     """Test marking case analyses as deleted."""
     # GIVEN case id for a case with analyses that are not deleted
-    analyses: Optional[List[Analysis]] = analysis_store.get_analyses_for_case(
+    analyses: Optional[list[Analysis]] = analysis_store.get_analyses_for_case(
```
Comment (diitaz93) [inline tests/store/crud/test_update.py:183]:
```suggestion
    analyses: list[Analysis] | None = analysis_store.get_analyses_for_case(
```

File: tests/store/crud/test_update.py [hunk 6]
```diff
@@ -180,15 +180,15 @@ def test_update_analysis_jobs_from_slurm_jobs(analysis_store: MockStore, squeue_
...
         case_id=ongoing_analysis_case_id
     )
     for analysis in analyses:
         assert not analysis.is_deleted
 
     # WHEN marking analyses as deleted
     analysis_store.update_case_analyses_as_deleted(case_id=ongoing_analysis_case_id)
-    analyses: Optional[List[Analysis]] = analysis_store.get_analyses_for_case(
+    analyses: Optional[list[Analysis]] = analysis_store.get_analyses_for_case(
```
Comment (diitaz93) [inline tests/store/crud/test_update.py:191]:
```suggestion
    analyses: list[Analysis] | None = analysis_store.get_analyses_for_case(
```

File: tests/store/crud/test_update.py [hunk 7]
```diff
@@ -204,7 +204,7 @@ def test_update_case_analyses_as_deleted_with_non_existing_case(
     # GIVEN case id for that do not exist
 
     # WHEN marking analyses as deleted
-    analyses: Optional[List[Analysis]] = analysis_store.update_case_analyses_as_deleted(
+    analyses: Optional[list[Analysis]] = analysis_store.update_case_analyses_as_deleted(
```
Comment (diitaz93) [inline tests/store/crud/test_update.py:207]:
```suggestion
    analyses: list[Analysis] | None = analysis_store.update_case_analyses_as_deleted(
```

File: trailblazer/apps/slurm/api.py [hunk 8]
```diff
@@ -14,31 +14,31 @@
...
+def _get_squeue_jobs_flag_input(slurm_job_id_file_content: dict[str, list[str]]) -> str:
     """Return a string of comma separated SLURM job ids to be used as input for squeue jobs flag."""
-    job_ids: List[str] = []
+    job_ids: list[str] = []
     for slurm_job_ids in slurm_job_id_file_content.values():
         [job_ids.append(str(job_id)) for job_id in slurm_job_ids]
     return ",".join(job_ids)
 
 
 def cancel_slurm_job(slurm_id: int, analysis_host: Optional[str] = None) -> None:
```
Comment (diitaz93) [inline trailblazer/apps/slurm/api.py:25]:
```suggestion
def cancel_slurm_job(slurm_id: int, analysis_host: str | None = None) -> None:
```

File: trailblazer/apps/slurm/api.py [hunk 9]
```diff
@@ -14,31 +14,31 @@
...
     """Cancel SLURM job by SLURM job id."""
-    scancel_commands: List[str] = ["scancel", str(slurm_id)]
+    scancel_commands: list[str] = ["scancel", str(slurm_id)]
     if analysis_host:
         scancel_commands = ["ssh", analysis_host] + scancel_commands
     subprocess.Popen(scancel_commands)
 
 
 def get_slurm_squeue_output(slurm_job_id_file: Path, analysis_host: Optional[str] = None) -> str:
```
Comment (diitaz93) [inline trailblazer/apps/slurm/api.py:33]:
```suggestion
def get_slurm_squeue_output(slurm_job_id_file: Path, analysis_host: str | None = None) -> str:
```

File: trailblazer/apps/slurm/api.py [hunk 10]
```diff
@@ -94,23 +94,23 @@ def _get_analysis_single_status(jobs_status_distribution: Dict[str, float]) -> s
...
     return any(
         ongoing_status in jobs_status_distribution
         for ongoing_status in TrailblazerStatus.ongoing_statuses()
     )
 
 
-def get_current_analysis_status(jobs_status_distribution: Dict[str, float]) -> str:
+def get_current_analysis_status(jobs_status_distribution: dict[str, float]) -> str:
     """Return current analysis status based on jobs status distribution."""
     single_analysis_status: Optional[str] = _get_analysis_single_status(
```
Comment (diitaz93) [inline trailblazer/apps/slurm/api.py:115]:
```suggestion
    single_analysis_status: str | None = _get_analysis_single_status(
```

File: trailblazer/apps/tower/models.py [hunk 11]
```diff
@@ -29,7 +29,7 @@ class TowerTask(BaseModel):
     tag: Optional[str] = None
     submit: Optional[str] = None
     complete: Optional[str] = None
```
Comment (diitaz93) [inline trailblazer/apps/tower/models.py:31]:
```suggestion
    tag: str | None = None
    submit: str | None = None
    complete: str | None = None
```

File: trailblazer/apps/tower/models.py [hunk 12]
```diff
@@ -29,7 +29,7 @@ class TowerTask(BaseModel):
...
-    module: List[str]
+    module: list[str]
     container: Optional[str] = None
     attempt: Optional[int] = None
     script: Optional[str] = None
```
Comment (diitaz93) [inline trailblazer/apps/tower/models.py:35]:
```suggestion
    container: str | None = None
    attempt: str | None = None
    script: str | None = None
```

**Review by diitaz93 — APPROVED** [review]
Great 👍 I realized my suggestions have more to do with [PEP 604](https://peps.python.org/pep-0604/) than with PEP 585. I think it would be nice to include them, but to fix all occurrences would extend the PR too much, so I understand if you want to disregard them

File: tests/apps/slurm/test_api.py (hunk 1 above)
Comment (henrikstranneheim) [reply tests/apps/slurm/test_api.py:1]: Let's do that PEP in another PR.

**Comment by henrikstranneheim** [comment]
### CLI and API 

```
θ64° [henrik.stranneheim@hasta:~] [S_base] 2m23s $ trailblazer scan
Error updating analysis for: case - grandaphid : PermissionError
Error updating analysis for: case - goodbengal : MissingSqueueOutput
Error updating analysis for: case - helpedlab : MissingSqueueOutput
Error updating analysis for: case - newgopher : PermissionError
Error updating analysis for: case - onestag : PermissionError
Error updating analysis for: case - funnyaphid : FileNotFoundError
Error updating analysis for: case - proudfish : PermissionError
Error updating analysis for: case - setcolt : PermissionError
Error updating analysis for: case - handyshrimp : FileNotFoundError
Error updating analysis for: case - divinetomcat : PermissionError
Error updating analysis for: case - HW2V2DRX2 : FileNotFoundError
Error updating analysis for: case - savingstag : FileNotFoundError
Error updating analysis for: case - topewe : PermissionError
Error updating analysis for: case - fullgoat : PermissionError
Error updating analysis for: case - bolddrum : PermissionError
Error updating analysis for: case - abovehusky : PermissionError
Error updating analysis for: case - chiefgannet : PermissionError
Error updating analysis for: case - grandoyster : PermissionError
Status in Tower: coolkodiak - 282044 - WPjrjs23yxQwq
Error logging case - coolkodiak :  JSONDecodeError
Error updating analysis for: case - setamoeba : FileNotFoundError
Error updating analysis for: case - unitedbeagle : FileNotFoundError
Error updating analysis for: case - equalbug : FileNotFoundError
Error updating analysis for: case - uphippo : FileNotFoundError
Status in Tower: truepigeon - 282656 - 1OGx9sb0sAuwRV
Error logging case - truepigeon :  ValidationError
Status in Tower: hardygrouse - 282660 - EgleejD4t4cNK
Status in Tower: fluentgull - 282661 - 26fRfyx4OgMLkn
Error logging case - fluentgull :  ValidationError
Updated status acetuna_rsync - 282662: failed
Status in Tower: tendergoose - 282664 - 2JjbLARhvk0cF9
All analyses updated!
```
Screenshot 2023-11-21 at 08 16 45

**Comment by henrikstranneheim** [comment]
Deployed to production:

```
Log deploy... done.
trailblazer, version 19.0.11
```

**Pull request merged**